File: tests/cuda/lua/test_signal_waveforms.lua
-- Signal Waveforms CUDA binding tests.
-- Run with:
--   LUA_PATH="bindings/lua/?/init.lua;;" LUA_CPATH="build/bindings/lua/?.so;;" \
--   LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda \
--   ~/.luarocks/bin/busted tests/cuda/lua/test_signal_waveforms.lua

local ok_gpu = pcall(function()
  local native = require("_insight")
  native.init()
end)
if not ok_gpu then
  print("SKIP: GPU backend not available")
  os.exit(0)
end

describe("Signal Waveforms CUDA Tests", function()
  local ins
  setup(function()
    ins = require("insight")
  end)

  -- ========================================================================
  -- Sawtooth (3 tests)
  -- ========================================================================

  it("sawtooth_basic", function()
    local t = ins.from_table({ 0.0, math.pi, 2.0 * math.pi - 1e-10 })
    local y = ins.signal.sawtooth(t, 1.0)
    assert.is_not_nil(y)
    assert.are.equal(3, y.numel)
    assert.near(-1.0, ins.item(y, 0), 1e-6)
    assert.near(0.0, ins.item(y, 1), 1e-6)
    assert.near(1.0, ins.item(y, 2), 1e-6)
  end)

  it("sawtooth_triangle", function()
    local t = ins.from_table({ 0.0, math.pi / 2.0, math.pi })
    local y = ins.signal.sawtooth(t, 0.5)
    assert.is_not_nil(y)
    assert.are.equal(3, y.numel)
    assert.near(-1.0, ins.item(y, 0), 1e-6)
    assert.near(0.0, ins.item(y, 1), 1e-6)
    assert.near(1.0, ins.item(y, 2), 1e-6)
  end)

  it("sawtooth_periodicity", function()
    local t = ins.from_table({ 0.5, 0.5 + 2.0 * math.pi, 0.5 + 4.0 * math.pi })
    local y = ins.signal.sawtooth(t, 1.0)
    assert.is_not_nil(y)
    assert.are.equal(3, y.numel)
    assert.near(ins.item(y, 0), ins.item(y, 1), 1e-6)
    assert.near(ins.item(y, 0), ins.item(y, 2), 1e-6)
  end)

  -- ========================================================================
  -- Square (2 tests)
  -- ========================================================================

  it("square_basic", function()
    local t = ins.from_table({ 0.0, math.pi, 2.0 * math.pi - 1e-10 })
    local y = ins.signal.square(t, 0.5)
    assert.is_not_nil(y)
    assert.are.equal(3, y.numel)
    assert.near(1.0, ins.item(y, 0), 1e-6)
    assert.near(-1.0, ins.item(y, 1), 1e-6)
    assert.near(-1.0, ins.item(y, 2), 1e-6)
  end)

  it("square_periodicity", function()
    local t = ins.from_table({ 1.0, 1.0 + 2.0 * math.pi, 1.0 + 4.0 * math.pi })
    local y = ins.signal.square(t, 0.5)
    assert.is_not_nil(y)
    assert.are.equal(3, y.numel)
    assert.near(ins.item(y, 0), ins.item(y, 1), 1e-6)
    assert.near(ins.item(y, 0), ins.item(y, 2), 1e-6)
  end)

  -- ========================================================================
  -- Gausspulse (2 tests)
  -- ========================================================================

  it("gausspulse_peak", function()
    local t = ins.from_table({ 0.0 })
    local y = ins.signal.gausspulse(t, 1000.0, 0.5)
    assert.is_not_nil(y)
    assert.are.equal(1, y.numel)
    assert.near(1.0, ins.item(y, 0), 1e-6)
  end)

  it("gausspulse_decay", function()
    local t = ins.from_table({ 0.0, 0.001, 0.01 })
    local y = ins.signal.gausspulse(t, 1000.0, 0.5)
    assert.is_not_nil(y)
    assert.are.equal(3, y.numel)
    assert.near(1.0, ins.item(y, 0), 1e-6)
    -- Decay: |y[0]| > |y[2]|
    assert.is_true(math.abs(ins.item(y, 0)) > math.abs(ins.item(y, 2)))
  end)

  -- ========================================================================
  -- Chirp (2 tests)
  -- ========================================================================

  it("chirp_linear", function()
    local t = ins.from_table({ 0.0 })
    local y = ins.signal.chirp(t, 6.0, 1.0, 10.0, "linear", 0)
    assert.is_not_nil(y)
    assert.are.equal(1, y.numel)
    assert.near(1.0, ins.item(y, 0), 1e-6)
  end)

  it("chirp_with_phase", function()
    local t = ins.from_table({ 0.0 })
    local y = ins.signal.chirp(t, 6.0, 1.0, 10.0, "linear", math.pi / 2.0)
    assert.is_not_nil(y)
    assert.are.equal(1, y.numel)
    assert.near(0.0, ins.item(y, 0), 1e-6)
  end)

  -- ========================================================================
  -- Unit Impulse (1 test)
  -- ========================================================================

  it("unit_impulse_center", function()
    local y = ins.signal.unit_impulse({ 11 })
    assert.is_not_nil(y)
    assert.are.equal(11, y.numel)
    assert.near(1.0, ins.item(y, 5), 1e-10)
    assert.near(0.0, ins.item(y, 0), 1e-10)
    assert.near(0.0, ins.item(y, 10), 1e-10)
  end)
end)

File: tests/cuda/python/test_signal_waveforms.py
"""Signal Waveforms CUDA binding tests.

Run with:
    PYTHONPATH=bindings/python:build/bindings/python \
    LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda \
    python -m pytest tests/cuda/python/test_signal_waveforms.py -v
"""

import sys
import os
import pytest
import math

_root = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..", "..")
sys.path.insert(0, os.path.join(_root, "bindings", "python"))
sys.path.insert(0, os.path.join(_root, "build", "bindings", "python"))

try:
    import insight as ins
    import numpy as np
except ImportError:
    pytest.skip("Insight or NumPy not available", allow_module_level=True)
try:
    ins.init()
    if ins.device_count() == 0:
        pytest.skip("No GPU available", allow_module_level=True)
except Exception:
    pytest.skip("GPU backend not available", allow_module_level=True)


class TestSawtooth:
    """Sawtooth waveform — tests 1-3."""

    def test_sawtooth_basic(self):
        t_np = np.array([0.0, math.pi, 2 * math.pi - 1e-10], dtype=np.float64)
        t = ins.from_numpy(t_np)
        y = ins.signal.sawtooth(t, width=1.0)
        d = y.numpy()
        assert abs(d[0] - (-1.0)) < 1e-6
        assert abs(d[1] - 0.0) < 1e-6
        assert abs(d[2] - 1.0) < 1e-6

    def test_sawtooth_triangle(self):
        t_np = np.array([0.0, math.pi / 2, math.pi], dtype=np.float64)
        t = ins.from_numpy(t_np)
        y = ins.signal.sawtooth(t, width=0.5)
        d = y.numpy()
        assert abs(d[0] - (-1.0)) < 1e-6
        assert abs(d[1] - 0.0) < 1e-6
        assert abs(d[2] - 1.0) < 1e-6

    def test_sawtooth_periodicity(self):
        t_np = np.array([0.5, 0.5 + 2 * math.pi, 0.5 + 4 * math.pi], dtype=np.float64)
        t = ins.from_numpy(t_np)
        y = ins.signal.sawtooth(t, width=1.0)
        d = y.numpy()
        assert abs(d[0] - d[1]) < 1e-6
        assert abs(d[0] - d[2]) < 1e-6


class TestSquare:
    """Square waveform — tests 4-5."""

    def test_square_basic(self):
        t_np = np.array([0.0, math.pi, 2 * math.pi - 1e-10], dtype=np.float64)
        t = ins.from_numpy(t_np)
        y = ins.signal.square_wf(t, duty=0.5)
        d = y.numpy()
        assert abs(d[0] - 1.0) < 1e-6
        assert abs(d[1] - (-1.0)) < 1e-6
        assert abs(d[2] - (-1.0)) < 1e-6

    def test_square_periodicity(self):
        t_np = np.array([1.0, 1.0 + 2 * math.pi, 1.0 + 4 * math.pi], dtype=np.float64)
        t = ins.from_numpy(t_np)
        y = ins.signal.square_wf(t, duty=0.5)
        d = y.numpy()
        assert abs(d[0] - d[1]) < 1e-6
        assert abs(d[0] - d[2]) < 1e-6


class TestGausspulse:
    """Gaussian pulse — tests 6-7."""

    def test_gausspulse_peak(self):
        t = ins.from_numpy(np.array([0.0], dtype=np.float64))
        y = ins.signal.gausspulse(t, fc=1000.0)
        assert abs(y.numpy()[0] - 1.0) < 1e-6

    def test_gausspulse_decay(self):
        t = ins.from_numpy(np.array([0.0, 0.001, 0.01], dtype=np.float64))
        y = ins.signal.gausspulse(t, fc=1000.0, bw=0.5)
        d = y.numpy()
        assert abs(d[0] - 1.0) < 1e-6  # peak at t=0
        assert abs(d[0]) > abs(d[2])  # decay


class TestChirp:
    """Chirp signal — tests 8-9."""

    def test_chirp_linear(self):
        t = ins.from_numpy(np.array([0.0], dtype=np.float64))
        y = ins.signal.chirp(t, f0=6.0, t1=1.0, f1=10.0, method="linear")
        assert abs(y.numpy()[0] - 1.0) < 1e-6

    def test_chirp_with_phase(self):
        # cos(0 + pi/2) = 0
        t = ins.from_numpy(np.array([0.0], dtype=np.float64))
        y = ins.signal.chirp(t, f0=6.0, t1=1.0, f1=10.0, method="linear", phi=math.pi / 2)
        assert abs(y.numpy()[0]) < 1e-6


class TestUnitImpulse:
    """Unit impulse — test 10."""

    def test_unit_impulse_center(self):
        y = ins.signal.unit_impulse(ins.Shape([11]))
        assert y.numel == 11
        d = y.numpy()
        for i in range(11):
            if i == 5:
                assert abs(d[i] - 1.0) < 1e-10
            else:
                assert abs(d[i]) < 1e-10


if __name__ == "__main__":
    pytest.main([__file__, "-v"])
